RecoParticleFlow/PFClusterProducer/plugins/alpaka/PFClusterSoAProducerKernel.dev.cc

```diff
@@ -0,0 +1,1634 @@
+      else if (pfRecHits[i].layer() == PFLayer::HCAL_ENDCAP)
+        rhENormInv = pfClusParams.recHitEnergyNormInvEE_vec()[pfRecHits[i].depth() - 1];
+      else {
+        rhENormInv = 0.;
+        printf("Rechit %d has invalid layer %d!\n", i, pfRecHits[i].layer());
+      }
+
+      iter = 0;
+      notDone = true;
+    }
+    alpaka::syncBlockThreads(acc);  // all threads call sync
+
+    int j = -1;  // j is the rechit index for this thread
+    int rhFracOffset = -1;
+    Position4 rhPos;
+    float rhEnergy = -1., rhPosNorm = -1.;
+
+    if (tid < nRHOther) {
+      rhFracOffset =
+          pfClusteringVars[i].seedFracOffsets() + tid + 1;  // Offset for this rechit in pcrhfrac, pcrhfracidx arrays
+      j = fracView[rhFracOffset].pfrhIdx();                 // rechit index for this thread
+      rhPos = Position4{pfRecHits[j].x(), pfRecHits[j].y(), pfRecHits[j].z(), 1.};
+      rhEnergy = pfRecHits[j].energy();
+      rhPosNorm = fmaxf(0., logf(rhEnergy * rhENormInv));
+    }
+    alpaka::syncBlockThreads(acc);  // all threads call sync
+
+    do {
+      if constexpr (debug) {
+        if (once_per_block(acc))
+          printf("\n--- Now on iter %d for topoId %d ---\n", iter, topoId);
+      }
+      float dist2 = -1., d2 = -1., fraction = -1.;
+      if (tid < nRHOther) {
+        // Rechit distance calculation
+        dist2 = (clusterPos.x - rhPos.x) * (clusterPos.x - rhPos.x) +
+                (clusterPos.y - rhPos.y) * (clusterPos.y - rhPos.y) +
+                (clusterPos.z - rhPos.z) * (clusterPos.z - rhPos.z);
+
+        d2 = dist2 / pfClusParams.showerSigma2();
+        fraction = clusterEnergy * rhENormInv * expf(-0.5 * d2);
+
+        // For single seed clusters, rechit fraction is either 1 (100%) or -1 (not included)
+        if (fraction > pfClusParams.minFracTot() && d2 < 100.)
+          fraction = 1.;
+        else
+          fraction = -1.;
+        fracView[rhFracOffset].frac() = fraction;
+      }
+      alpaka::syncBlockThreads(acc);  // all threads call sync
+
+      if constexpr (debug) {
+        if (once_per_block(acc))
+          printf("Computing cluster position for topoId %d\n", topoId);
+      }
+
+      if (once_per_block(acc)) {
+        // Reset cluster position and energy
+        clusterPos = seedPos;
+        clusterEnergy = seedEnergy;
+      }
+      alpaka::syncBlockThreads(acc);  // all threads call sync
+
+      // Recalculate cluster position and energy
+      if (fraction > -0.5) {
+        alpaka::atomicAdd(acc, &clusterEnergy, rhEnergy, alpaka::hierarchy::Blocks{});
+        alpaka::atomicAdd(acc, &clusterPos.x, rhPos.x * rhPosNorm, alpaka::hierarchy::Blocks{});
+        alpaka::atomicAdd(acc, &clusterPos.y, rhPos.y * rhPosNorm, alpaka::hierarchy::Blocks{});
+        alpaka::atomicAdd(acc, &clusterPos.z, rhPos.z * rhPosNorm, alpaka::hierarchy::Blocks{});
+        alpaka::atomicAdd(acc, &clusterPos.w, rhPosNorm, alpaka::hierarchy::Blocks{});  // position_norm
```

Comment:
`alpaka::hierarchy::Blocks{}` means the the operations is required to be atomic across different blocks of the same kernel grid (I know, it's confusing, sorry :-().

In this case, the variables are local to the block, so the operations are required to be atomic across different threads of the same block. This can be written using `alpaka::hierarchy::Threads{}`:

```suggestion
        alpaka::atomicAdd(acc, &clusterEnergy, rhEnergy, alpaka::hierarchy::Threads{});
        alpaka::atomicAdd(acc, &clusterPos.x, rhPos.x * rhPosNorm, alpaka::hierarchy::Threads{});
        alpaka::atomicAdd(acc, &clusterPos.y, rhPos.y * rhPosNorm, alpaka::hierarchy::Threads{});
        alpaka::atomicAdd(acc, &clusterPos.z, rhPos.z * rhPosNorm, alpaka::hierarchy::Threads{});
        alpaka::atomicAdd(acc, &clusterPos.w, rhPosNorm, alpaka::hierarchy::Threads{});  // position_norm
```